Blog post: From React to Ember - Building a Personal Website/Blog (#2057)
mainmatter/mainmatter.com: src/posts/2023-05-28-from-react-to-ember-building-a-blog.md

world progressed, I knew it was time to build my own website.

However, the amount of different tools for creating a website and the fact that
I never have done one on my own was overwhelming me. I shared these dramas with
my working colleague, Chris. From that day forward, we began
scheduling weekly pair programming sessions. Chris became my accountability
partner and mentor, bringing immense value to my journey. 💜

Working at a consultancy company like Mainmatter means that you don't always
collaborate with everyone on the team. Chris and I never had the opportunity to
work together on a client project. However, I believe it's fantastic to have the
chance to collaborate with someone outside of your usual circle. As the least
experienced member of the company, these sessions have proven to be incredibly
valuable for me.

![Pair programming session with Chris running lighthouse to check my website performace. ](/assets/images/posts/2023-05-28-from-react-to-ember-building-a-blog/calls-with-chris.png)_Chris
and I in one of our pair programming sessions. This one was about improving the
performance of my website, so we were running lighthouse. Spoiler alert - not

> rendering efficently my photos was my big issue._

Reviewer: ```suggestion
rendering my photos efficiently was my big issue._
```

import { tracked } from "@glimmer/tracking";

export default class Dropdown extends Component {
  @tracked open = false;

  @action
  dropdown() {
    this.open = !this.open;
  }
}
```

```handlebars
{% raw %} // header.hbs
<Dropdown @title="Blog">
  <LinkTo class="dropdown-text" @route="posts"> All Posts</LinkTo>
  <LinkTo class="dropdown-text" @route="posts.category" @model="travel">
    Travel</LinkTo>
  <LinkTo class="dropdown-text" @route="posts.category" @model="tech">
    Tech</LinkTo>
</Dropdown>
{% endraw %}
```

I mean, from someone coming from React, there was a lot of new stuff going on in
these two files.

Also, going back to using `this.` keyword made me relive my university times and
how understanding it was the boss level of object-oriented programming. 😂

Let me break down what is happening in my `<Dropdown>`:

_On a side note, I want to say that this is probably not the most helpful
dropdown component since I can only customize its title. In my case, I ended up
not reusing it, so I didn’t feel the need to make it more customizable._

- My dropdown component has a button that renders a
  {% raw %}`{{@title}}`{% endraw %}. In React, this would be the `props.title`.

> - Then I have a `<div>` that will receive {% raw %}`{{yield}}`{% endraw %}. This
>   was a new keyword for me as well. This was a new keyword for me as well. The

Reviewer: ```suggestion
- Then I have a `<div>` that will receive {% raw %}`{{yield}}`{% endraw %}. This was a new keyword for me as well. The
```
Just a duplicate sentence 👍

  the “children” property.
- In here
  {% raw %}`class='dropdown-button {{if this.open "focus"}}'`{% endraw %}, I’m
  dynamically changing the button class name so that when the open variable is
  set to true, I add ‘focus’ to the class name, which is how I apply the yellow
  color to the dropdown title when the user opens it. Same happens with
  {% raw %}`<div class='dropdown-list {{if this.open "show"}}'>`{% endraw %}.
- Next we have {% raw %}`{{on 'click' this.dropdown}}`{% endraw %} where I'm
  saying to my component that when someone clicks it, run the
  `dropdown()`function - defined in the _dropdown.js_ file.
- Time to dive into _dropdown.js_ file associated to my component:
  - Pretty much all my components have these 3 imports in their javascript
    files:
    - `Component` is a class from the Glimmer library (DOM rendering engine
      made by the ember team) that allows me to define the behavior of my
      component and its UI logic. In the latest versions of React, functional components were introduced, meaning that they can be defined using regular JavaScript functions - this way, there is no need to extend a class from a Component base class.

>     - `{action}` is a method coming from the `@ember/object` module. It is Ember's own

Reviewer: this is a decorator so I think it might be good to call it a decorator instead of a method 🤔 

```suggestion
    - `{action}` is a decorator coming from the `@ember/object` module. It is Ember's own
```

      object model, it allows me to define an event handler, in this case, the
      `dropdown()` function that gets triggered when the user clicks on the

>       button. In React, we can use regular functions as event handlers directly.

Reviewer: So this is a bit of a tricky one. You use regular functions in ember too without any issue, the only reason we need to use `@action` in this context is because we are using `this` in the function and it needs to bind the Glimmer Template context to the context of the class. Nowadays you can think of `@action` in the same way you would use `bind` in Javascript but I don't know if that is a helpful analogy 😂

Author: Ok, I added more detail to this explanation. Can you tell me if you think this new explanation is correct?

      changes, changes in the UI will be triggered. In React, we typically
      achieve this functionality using the built-in state management and the
     `useState` hook. In my website this translates to - whenever a user clicks
      on the button, the value of the tracked variable changes, which controls
      the visibility of the dropdown's children and applies specific CSS
      classes.
- Finally, it's time to look at the last file - the template where I use my
  `<Dropdown>` component, **header.hbs.**:
  - I’m passing to the `<Dropdown>` a list of `<LinkTo>`. The latter component
    allows me to create a link to a route, the `@model` argument I use, allows
    my app to know which data to retrieve when clicking on that route.

### Rendering My Blogposts

1. **Finding a tool to publish my posts**

   After giving the shape I wanted for my website, the question for one of the
   key features was - ‘How do I create a not-so-complicated system that allows
   me to publish blog posts?’ ‘Am I going to create HTML pages every time?´

   Chris proposed an approach - ‘Let’s create posts using markdown files.’ He
   said we could use this tool - `ember-cli-showdown`.

>    This addon converts my markdown into readable HTLM objects, so I don’t need

Reviewer: ```suggestion
   This addon converts my markdown into HTML, so I don’t need
```

   to have that tedious & time-consuming job of writing an HTML file every time
   I want to publish a new post. It was a deal. 🤝

2. **Combining my posts information with my tool**

   _'How can I extract what I write in my markdown files into my HTML
   templates?'_ Another plugin enters the room: welcome to
   [**Broccoli-Static-Site-Json.**](https://www.npmjs.com/package/broccoli-static-site-json)
   👋🏼 It generates a JSON API of the content of my markdowns. Let me guide you
   through the process:

   - I need to define what data will populate my JSON API, to do it, I define a
     ‘Post’ model. It looks like this:

   ```js
   //app/models/post.js
   import Model, { attr } from "@ember-data/model";
   export default class PostModel extends Model {
     @attr title;
     @attr("date") date;
     @attr category;
     @attr description;
     @attr image;
     @attr content;
     @attr header;
   }
   ```

>    Now that I have them defined, I need to tell broccoli addon what attributes

Reviewer: ```suggestion
   Now that I have them defined, I need to tell the broccoli addon what attributes
```

   to use when building the JSON API.

>    Let’s add that in `/lib/content-generator/index.js` (a file that the addon
>    adds it to your project):

Reviewer: ```suggestion
   Let’s add that in `/lib/content-generator/index.js` (an in-repo Ember addon that you can use to add broccoli-static-site-json to your project):
```
I kinda feel bad suggesting this 🙈 it's a bit more correct but probably a bit too much detail

   ```js
   //lib/content-generator/index.js
   const postsTree = new StaticSiteJson("data/post", {
     type: "post",
     contentFolder: "data/posts",
     attributes: ["title", "date", "category", "image", "description"],
     collate: true,
   });
   ```

   I start by specifying the location of the API information, which, in my case,
   is stored in my markdown files located under 'data/post'.

   The 'type' parameter defines the name of the JSON:API document. The name you
   assign to the 'type' parameter gets pluralized, so for the next attribute, it
   should be 'data/posts'. Next, I define the attributes I want to include.

>    Additionally, I utilize the last configuration variable, which allows me to

Reviewer: ```suggestion
   Additionally, I utilize the last configuration variable, `collate`, which allows me to
```

can become messy; in Ember, you don't need to care about it. In Ember, having
all these separate files that target the same page can confuse me - controllers,
.js, .hbs. However, some people prefer it. In the end, there is also a
preference matter.

Ember has a more traditional approach to object-oriented programming. It
resembled what I learned at university way more.

Regarding the type of project I built - a personal website/blog. I might be
falling into the group of people who create a blog and never post on it. 🙈 I
think I put a lot of pressure when I start working on a post, and I made them
very long - this adds friction to posting regularly. Also, when I designed how
to present the information, I just kept long posts in mind.

My pieces of advice are: to choose a publishing tool that is easy enough - the
markdown files approach is working well for me; when sketching your website,
keep in mind short posts as well.

Since almost every human is an internet user nowadays, it is nice to have your
website where you can share your work and interests - you can end up meeting
more people with the same interests as you and possibly cool work
collaborations.

> If you want to check my internet corner you can access

Reviewer: ```suggestion
If you want to check my internet corner you can access it 
```

I got myself a domain at [namecheap](https://www.namecheap.com/) and I used
[netlify](https://www.netlify.com/) to deploy it.

I faced some challenges with the deployment:

1. Image fingerprinting The photos were not loading on the deployed website.
   Why? The browser was adding a hash to my photo sources, e.g - from
   ‘work-in-cv.jpg’ it would become ‘work-in-cv-h4323524hgs.jpg’. In my HTML the
   photo source would still be the one without the hash, but that path would not
   exist anymore. How did I fix it? I specified on ember-cli-build what type of
   files I wanted to get fingerprinted:

```js
//ember-cli-build.js
module.exports = function (defaults) {
  let app = new EmberApp(defaults, {
    // Add options here
    fingerprint: {
      extensions: ['js', 'css', 'map'], //remove photo formats from here
    },
  }
}
```

2. Routes were not found

Ember is a SPA (Single Page Application), meaning that the browser just
recognizes the index page and only knows how to go to the other pages from
there, but if I would write down a specific URL, it won't know how to handle it.

Two ways to fix it:

1. Creating HTML files in the paths of the routes.

> 2. If we want to keep this a SPA - we tell Netlify to give us the index.html

Author: here I meant compared to the 1st solution (that doesn't follow a SPA approach). If we want to keep SPA logic, we opt for the second way. I will drop it, it is not important and I guess it is not clear.

```handlebars
{% raw %} // dropdown.hbs
<div class="dropdown">
  <button
    class="dropdown-button {{if this.open 'focus'}}"
    type="button"
    {{on "click" this.dropdown}}
  >
    {{@title}}
    <img class="dropdown-icon {{if this.open 'focus'}}" />
  </button>
  <div class="dropdown-list {{if this.open 'show'}}">
    {{yield}}
  </div>
</div>
{% endraw %}
```

```js
//dropdown.js
import Component from "@glimmer/component";
import { action } from "@ember/object";
import { tracked } from "@glimmer/tracking";

export default class Dropdown extends Component {
  @tracked open = false;

  @action
  dropdown() {
    this.open = !this.open;
  }
}
```

```handlebars
{% raw %} // header.hbs
<Dropdown @title="Blog">
  <LinkTo class="dropdown-text" @route="posts"> All Posts</LinkTo>
  <LinkTo class="dropdown-text" @route="posts.category" @model="travel">
    Travel</LinkTo>
  <LinkTo class="dropdown-text" @route="posts.category" @model="tech">
    Tech</LinkTo>
</Dropdown>
{% endraw %}
```

I mean, from someone coming from React, there was a lot of new stuff going on in
these two files.

Also, going back to using `this.` keyword made me relive my university times and

> how understanding it was the boss level of object-oriented programming. 😂

Author: This will change what I want to express in this sentence. I mean that understanding the 'this' keyword was the boss level. Is it confusing?  I can rephrase it a bit to get more clear.             
"Also, going back to using `this.` keyword brought memories from my university days and
how understanding this concept was the boss level of object-oriented programming." 😂

not reusing it, so I didn’t feel the need to make it more customizable._

- My dropdown component has a button that renders a
  {% raw %}`{{@title}}`{% endraw %}. In React, this would be the `props.title`.
- Then I have a `<div>` that will receive {% raw %}`{{yield}}`{% endraw %}. This
  was a new keyword for me as well. The yield acts like a placeholder, and if I
  use `<Dropdown>` in another component, which is the case in the _header.bhs_,
  what I pass within the open and closing tags of my component will be rendered
  in the place of{% raw %}`{{yield}}`{% endraw %}. The equivalent of this in
  React would be the “children” property.
- In here
  {% raw %}`class='dropdown-button {{if this.open "focus"}}'`{% endraw %}, I’m
  dynamically changing the button class name so that when the open variable is
  set to true, I add ‘focus’ to the class name, which is how I apply the yellow
  color to the dropdown title when the user opens it. Same happens with
  {% raw %}`<div class='dropdown-list {{if this.open "show"}}'>`{% endraw %}.
- Next we have {% raw %}`{{on 'click' this.dropdown}}`{% endraw %} where I'm

>   saying to my component that when someone clicks it, run the

Author: I would like to keep the mention of the click event. 
If I rephrase to this:
Next, we have {% raw %}`{{on 'click' this.dropdown}}`{% endraw %}, this instructs my app to run the
  `dropdown()`function (defined in the _dropdown.js_ file) after the user clicks on the button.
  how does it sound?

>   `dropdown()`function - defined in the _dropdown.js_ file.

Author: I don't think this makes sense

- Time to dive into _dropdown.js_ file associated to my component:
  - Pretty much all my components have these 3 imports in their javascript
    files:
    - `Component` is a class from the Glimmer library (DOM rendering engine made
      by the ember team) that allows me to define the behavior of my component
      and its UI logic. In the latest versions of React, functional components
      were introduced, meaning that they can be defined using regular JavaScript
      functions - this way, there is no need to extend a class from a Component
      base class.
    - `{action}` is a decorator coming from the `@ember/object` module. It is
      Ember's own object model, it allows me to define an event handler, in this
      case, the `dropdown()` function that gets triggered when the user clicks
      on the button. `@action` here works similarly to `bind()` in JavaScript -
      when I use `this.open` inside of my Dropdown class at the dropdown.js
      file, I need to find a way to make sure `this` refers to the same context
      as the one used in the template (Dropdown.hbs file), and `@action` does
      this for us. This way, when calling the `dropdown()` function from the

>       template the right context (`this`) gets passed to it. files React doesn't

Reviewer 2: I think a piece is missing here after the "."

  return `posts/${file.replace(/\.md$/, "")}`;
});

module.exports = function (defaults) {
  let app = new EmberApp(defaults, {
    prember: {
      urls: ["/", "/about-me", "/contact", "/posts", ...posts],
    },
  });
  return app.toTree();
};
```

How do I keep Prember aware of the URLs of my posts (since I’ll keep adding new
ones)?

I built a function to help me with this, which you can see in the code above:

```js
//ember-cli-build.js
let posts = fs.readdirSync("data/post").map(file => {
  return `posts/${file.replace(/\.md$/, "")}`;
});
```

I create a posts array that is populated with all the files I have under
data/post directory. I make some changes before I save it to my array, the files
there look like this _working-from-cv.md_, I remove the _.md_ and I attach
_posts/_, so at the end I have an array of the URLs to my posts that look like
this _posts/working-from-cv_.

Now, I just need to add the array containing all the URLs to my posts with all
the other URLs for the other pages -
`urls: ['/', '/about-me', '/contact', '/posts', ...posts]`, with `…posts`,
javascript will deconstruct my array and add its values to the urls array.

## My Thoughts About This Project

> Seven months have passed since I released it.

Reviewer 2: ```suggestion
Seven months have passed since I released my new blog.
```

After I deployed it, I made some improvements, primarily related to the

> performance. After running Lighthouse, I saw that my biggest issue was not being

Reviewer 2: ```suggestion
performance. By running Lighthouse, I saw that my biggest issue was not being
```

efficient when loading my images. I recently added a little reaction button at
the end of my posts. (inspired by Josh Comeau's website).

I don't have a long experience in web development, so for me, going from react
to ember it wasn't the smoothest transition. I found Ember a bit harder. I guess

> every framework has its pros and cons. In React, dealing with component states

Reviewer 2: ```suggestion
every framework has its pros and cons. 
In React, dealing with component states
```

can become messy; in Ember, you don't need to care about it. In Ember, having

> all these separate files that target the same page can confuse me - controllers,

Reviewer 2: ```suggestion
all these separate files that target the same page can create confusion - controllers,
```

> .js, .hbs. However, some people prefer it. In the end, there is also a

Reviewer 2: ```suggestion
.js, .hbs. However, some people prefer this approach. In the end, choices like these are very subjective.
```

> preference matter.

Reviewer 2: ```suggestion
```

Ember has a more traditional approach to object-oriented programming. It
resembled what I learned at university way more.

> Regarding the type of project I built - a personal website/blog. I might be If

Reviewer 2: I am not sure what you mean with this conclusion

Author: I think some information got lost here 🙈 I will rephrase it.

you want to check my internet corner you can access it think I put a lot of
pressure when I start working on a post, and I made them very long - this adds
friction to posting regularly. Also, when I designed how to present the
information, I just kept long posts in mind.

> My pieces of advice are: to choose a publishing tool that is easy enough - the

Reviewer 2: ```suggestion
My pieces of advice are: choose a publishing tool that is easy enough - the
```

markdown files approach is working well for me; when sketching your website,

> keep in mind short posts as well.

Reviewer 2: ```suggestion
keep short posts in mind as well.
```

Since almost every human is an internet user nowadays, it is nice to have your

> website where you can share your work and interests - you can end up meeting

Reviewer 2: ```suggestion
own website where you can share your work and interests - you can end up meeting
```

> more people with the same interests as you and possibly cool work

Reviewer 2: ```suggestion
more people with the same interests as you and possibly initiate cool work
```

---

> title: "From React To Ember: Building a Personal Website/Blog"

Reviewer 2: ```suggestion
title: "From React To Ember: Building a Personal Website"
```
